File: dragging rectangles with mouse.js
// https://editor.p5js.org/
// Does the same as the Python code. 

let rect1, rect2;
let dragging = null;
let offsetX = 0;
let offsetY = 0;

function setup() {
  createCanvas(800, 600);
  
  // Määrittele suorakulmiot muodossa {x, y, w, h}
  rect1 = { x: 100, y: 100, w: 150, h: 100 };
  rect2 = { x: 300, y: 200, w: 150, h: 100 };
}

function draw() {
  background(255);

  // Piirrä suorakulmiot
  fill(255, 0, 0);
  rect(rect1.x, rect1.y, rect1.w, rect1.h);

  fill(0, 0, 255);
  rect(rect2.x, rect2.y, rect2.w, rect2.h);
}

function mousePressed() {
  if (isInside(mouseX, mouseY, rect1)) {
    dragging = rect1;
    offsetX = rect1.x - mouseX;
    offsetY = rect1.y - mouseY;
  } else if (isInside(mouseX, mouseY, rect2)) {
    dragging = rect2;
    offsetX = rect2.x - mouseX;
    offsetY = rect2.y - mouseY;
  }
}

function mouseReleased() {
  dragging = null;
}

function mouseDragged() {
  if (dragging !== null) {
    let oldX = dragging.x;
    let oldY = dragging.y;

    dragging.x = mouseX + offsetX;
    dragging.y = mouseY + offsetY;

    if (rectanglesOverlap(rect1, rect2)) {
      dragging.x = oldX;
      dragging.y = oldY;
    }
  }
}

// Apufunktiot
function isInside(x, y, r) {
  return x >= r.x && x <= r.x + r.w &&
         y >= r.y && y <= r.y + r.h;
}

function rectanglesOverlap(a, b) {
  return !(a.x + a.w <= b.x ||
           b.x + b.w <= a.x ||
           a.y + a.h <= b.y ||
           b.y + b.h <= a.y);
}

File: dragging rectangles with mouse.py
# pip install pygame, if needed. 

import pygame

# Alusta pygame
pygame.init()

# Ikkunan koko
WIDTH, HEIGHT = 800, 600
screen = pygame.display.set_mode((WIDTH, HEIGHT))
pygame.display.set_caption("Siirrä suorakulmioita hiirellä (ei päällekkäin)")

# Värit
WHITE = (255, 255, 255)
RED = (255, 0, 0)
BLUE = (0, 0, 255)

# Suorakulmiot (x, y, width, height)
rect1 = pygame.Rect(100, 100, 150, 100)
rect2 = pygame.Rect(300, 200, 150, 100)

# Muuttujat vetämiseen
dragging = None
offset_x = 0
offset_y = 0

# Pääsilmukka
running = True
clock = pygame.time.Clock()

def rectangles_overlap(r1, r2):
    return r1.colliderect(r2)

while running:
    screen.fill(WHITE)

    # Piirrä suorakulmiot
    pygame.draw.rect(screen, RED, rect1)
    pygame.draw.rect(screen, BLUE, rect2)

    for event in pygame.event.get():
        if event.type == pygame.QUIT:
            running = False

        elif event.type == pygame.MOUSEBUTTONDOWN:
            if rect1.collidepoint(event.pos):
                dragging = rect1
                offset_x = rect1.x - event.pos[0]
                offset_y = rect1.y - event.pos[1]
            elif rect2.collidepoint(event.pos):
                dragging = rect2
                offset_x = rect2.x - event.pos[0]
                offset_y = rect2.y - event.pos[1]

        elif event.type == pygame.MOUSEBUTTONUP:
            dragging = None

        elif event.type == pygame.MOUSEMOTION:
            if dragging:
                new_x = event.pos[0] + offset_x
                new_y = event.pos[1] + offset_y
                old_rect = dragging.copy()
                dragging.x = new_x
                dragging.y = new_y

                # Estetään päällekkäisyys siirron jälkeen
                if (dragging == rect1 and rectangles_overlap(rect1, rect2)) or \
                   (dragging == rect2 and rectangles_overlap(rect1, rect2)):
                    dragging.x = old_rect.x
                    dragging.y = old_rect.y

    pygame.display.flip()
    clock.tick(60)

pygame.quit()
